src/Plugin/tmgmt/Translator/ThebigwordTranslator.php

```diff
@@ -0,0 +1,872 @@
+      }
+      else {
+        $uri = '';
+      }
+      $state = 'ResourcePreviewUrl';
+      $name = 'preview-url';
+    }
+    $preview_data = '<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE PreviewUrl SYSTEM "http://www.thebigword.com/dtds/PreviewUrl.dtd">
+<PreviewUrl>' . $_SERVER['HTTP_HOST'] . '/' . $uri . '</PreviewUrl>';
+
+    /** @var \DateTime $required_by */
+    $required_by = $job_item->getJob()->getSetting('required_by');
+    $required_by->setTimezone(new \DateTimeZone('UTC'));
+    $datetime = $required_by->format('Y-m-d\TH:i:s');
+    $form_params = [
+      'ProjectId' => $project_id,
+      'RequiredByDateUtc' => $datetime,
+      'SourceLanguage' => $job_item->getJob()->getRemoteSourceLanguage(),
+      'TargetLanguage' => $job_item->getJob()->getRemoteTargetLanguage(),
+      'FilePathAndName' => "$name.xml",
+      'FileState' => $state,
+      'FileData' => base64_encode($preview_data),
+      'FileIdToUpdate' => $file_id,
+    ];
+    $file_id = $this->request('file', 'PUT', $form_params);
+
+    return $file_id;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pullRemoteTranslations(Translator $translator) {
+    $translated = 0;
+    $untranslated = 0;
+    $result = $this->pullRemoteTranslationsForStatus($translator, 'TranslatableReviewPreview');
+    $translated += $result['translated'];
+    $untranslated += $result['untranslated'];
+    $result = $this->pullRemoteTranslationsForStatus($translator, 'TranslatableComplete');
+    $translated += $result['translated'];
+    $untranslated += $result['untranslated'];
+    return [
+      'updated' => $translated,
+      'non-updated' => $untranslated,
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pullRemoteTranslationsForStatus(Translator $translator, $status) {
+    $this->setTranslator($translator);
+    $result = RemoteMapping::loadByLocalData(NULL, NULL, 'tmgmt_thebigword');
+    $remotes = [];
+    /** @var RemoteMapping $remote */
+    foreach ($result as $remote) {
+      $remotes[$remote->getRemoteIdentifier2()] = $remote;
+    }
+    $translated = 0;
+    $not_translated = 0;
+
+    try {
+      $files = $this->requestFileinfo($status);
+      foreach ($files as $file) {
+        $file_id = $file['FileId'];
+        $project_id = $file['ProjectId'];
+        /** @var RemoteMapping $remote */
+        $remote = isset($remotes[$project_id]) ? $remotes[$project_id] : NULL;
+        if ($remote != NULL && isset($remote->getRemoteData('files')[$file_id]) && $file['FileStateVersion'] == $remote->getRemoteData('files')[$file_id]['FileStateVersion']) {
+          $this->addFileDataToJob($remote->getJob(), $status, $file_id);
+          $translated++;
+        }
+        else {
+          $not_translated++;
+        }
+      }
+    }
+    catch (TMGMTException $e) {
+      drupal_set_message(new TranslatableMarkup('Could not pull translation resources due to the following error: @message', ['@message' => $e->getMessage()]), 'error');
```

Comment:
API functions should not call drupal_set_message(). That's what exceptions are for. make sure this is cached somewhere in a UI related method/function that knows that we have a user. Someone might have a cron job that calls this once per hour or something, who knows.